File: TP1/Q14/Arquivo.java
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class Arquivo {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int numLidos = sc.nextInt();

        // Consome a quebra de linha deixada pelo nextInt() para evitar problemas na leitura posterior
        sc.nextLine();

        // Tenta abrir (ou criar) um arquivo de acesso aleatório chamado "Doubles.txt" para leitura e escrita
        try (RandomAccessFile file = new RandomAccessFile("Doubles.txt", "rw")) {

            // Loop para ler os números da entrada e escrevê-los no arquivo
            for (int i = 0; i < numLidos; i++) {
                double num; 
                String linha = sc.nextLine(); 
                num = Double.parseDouble(linha);

                file.writeDouble(num);
            }

            // Loop para ler os números do arquivo em ordem inversa e exibi-los na tela
            for (int i = numLidos - 1; i >= 0; i--) {
                // Posiciona o ponteiro do arquivo no início do i-ésimo número double
                // Cada double ocupa 8 bytes, então o offset é i * 8
                file.seek(i * 8);
                double num = file.readDouble();

                if (num == (int) num) {
                    System.out.println((int) num);
                } else {
                    System.out.println(num); 
                }
            }

        } catch (IOException erro) {
            
        } finally {
          
            sc.close();
        }
    }
}

File: TP1/Q15/Arquivo.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
    int qtdnumeros;


    scanf("%i", &qtdnumeros);
    getchar(); // Consome a quebra de linha deixada pelo scanf

   
    FILE *file = fopen("Doubles.txt", "wb+");
    char linha[50];

    // Loop para ler os números e escrevê-los no arquivo
    for (int i = 0; i < qtdnumeros; i++) {
        
        fgets(linha, 50, stdin);
        linha[strcspn(linha, "\n")] = '\0';

        // Verifica se a linha começa com um ponto (.)
        if (linha[0] == '.') {
            
            for (int z = strlen(linha); z >= 0; z--) {
                linha[z + 1] = linha[z];
            }

            linha[0] = '0';
        }

        char *endptr;
        double num = strtod(linha, &endptr);

        fwrite(&num, sizeof(double), 1, file);
    }

    // Loop para ler os números do arquivo em ordem inversa
    for (int i = qtdnumeros - 1; i >= 0; i--) {
        double num;

        // Posiciona o ponteiro do arquivo no início do i-ésimo double
        fseek(file, i * sizeof(double), SEEK_SET);
        fread(&num, sizeof(double), 1, file);

        // Verifica se o número é um inteiro
        if (num == (int)num) {
           
            int n = (int)num;
            printf("%d\n", n);
        } else {
            printf("%g\n", num);
        }
    }

    // Fecha o arquivo
    fclose(file);

    return 0;
}